calyx/src/lang/context.rs

```diff
@@ -121,7 +149,21 @@ impl Context {
         }
     }
 
-    // XXX(sam) need a way to insert components
+    /// Insert the component `comp` into `self`.
+    pub fn insert_component(&self, comp: Component) {
+        // It's possible that this method will be called inside the
+        // `definitions_iter` function. In that case, the borrow will
```

Comment:
How can a borrow fail? Isn't it compile time?

Reply:
I'm talking about the `.borrow()` and `.borrow_mut()` methods on `RefCell`s. I'll make this more clear. In the code above, I can take the immutable `&self`. and use `self.definitions.borrow_mut()` to get mutable access to the hashmap. But because Rust thinks the reference to self is immutable, it will let multiple of them be taken. So with the use of `borrow_mut()` its possible to take multiple mutable references to the same data, which ofc violates Rust's safety guarantees. `RefCell` deals with this by checking the borrow rules at runtime rather than compile time. So the second `borrow_mut()` produces a runtime error. 

Consider the following toy code:
```rust
let mut x = 4;
let y = &mut x;
let z = &mut x; <---- compile time error!
``` 
If instead we do this:
```rust
let x = RefCell::new(4);
let y = x.borrow_mut();
let z = x.borrow_mut(); <---- safe at compile time, run time error!
```

Reply:
Oh weird. Didn't know about RefCells. Anyways, this looks good!